In [1]:
import pandas as pd

In [2]:
# load dataset

transactions = pd.read_csv("customer_transactions_contaminated.csv")

## Inspect and Clean Customer Transactions Dataset

In [3]:
# inspect dataset

print("First 20 rows: ")
display(transactions.head(20))

print("\nDataset Information: ")
display(transactions.info())

print("\nDataset Dimensions: ")
display(transactions.shape)

print("\nDataset Columns: ")
display(transactions.columns)

print("\nDataset Numerical Statistics: ")
display(transactions.describe())

First 20 rows: 


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
0,60567026-f719-4cd6-849e-137e86d8938f,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-05-15,117.64,Clothing,PayPal
1,4090ba85-b111-4f75-a792-c777965f5255,2c39b9fe-ff57-4d39-9321-9f5cdf187aa1,2023-04-26,466.14,Health & Beauty,Bank Transfer
2,9223891b-73ff-4d5c-b8ae-13ece82ee28b,f79588dd-3db9-4ffa-97f8-7de0e64259f1,2022-09-23,563.99,Clothing,Debit Card
3,9243eebc-938f-480c-8564-16d503d250de,401c0fc9-60df-4455-ad78-67c132f9897d,2024-04-15,254.44,Automotive,PayPal
4,6e3e8eb8-bc0f-4ffe-9f74-5d5efec9502f,2034aebc-8280-4254-a667-92bcd1c2be4f,2024-06-03,590.52,Home & Garden,Bank Transfer
5,3eccdcf9-e8ce-4e44-9f62-8bef88e70672,78dfb45e-16e0-4ecc-8f67-f2e210c31c3e,2024-04-07,Free,NaN,Credit Card
6,7fbc5847-a652-4fc8-a7d8-5e85d846e91b,591f1ea2-c97d-44dd-9539-71ac89b94ff3,2024-01-12,NaN,Automotive,PayPal
7,958e5c8d-48ca-42dd-bb71-a766a374233a,fb24e098-3ab9-40a2-bcc3-b8ebb23f549a,2023-03-10,399.7,Home & Garden,PayPal
8,39c6e7d2-6c4b-44c0-8961-ddc1ecbdb0c6,833b026a-7c02-4101-832d-62c07569b0f6,2024-01-26,296.99,Clothing,PayPal
9,474f9233-0616-431b-b3e9-a6feabe68abb,f8bde21a-7a6b-41f6-ac39-eeb77640fa9b,2023-06-15,149.39,Clothing,Debit Card



Dataset Information: 
<class 'pandas.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   CustomerID       3200 non-null   str  
 1   TransactionID    3200 non-null   str  
 2   TransactionDate  3200 non-null   str  
 3   Amount           2896 non-null   str  
 4   ProductCategory  2901 non-null   str  
 5   PaymentMethod    3200 non-null   str  
dtypes: str(6)
memory usage: 486.8 KB


None


Dataset Dimensions: 


(3200, 6)


Dataset Columns: 


Index(['CustomerID', 'TransactionID', 'TransactionDate', 'Amount',
       'ProductCategory', 'PaymentMethod'],
      dtype='str')


Dataset Numerical Statistics: 


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
count,3200,3200,3200,2896,2901,3200
unique,1871,3000,808,2603,5,4
top,49f06ea9-4ae5-4ce5-98ce-d2c2f6f15cbd,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-01-06,-100,Clothing,Debit Card
freq,7,2,13,39,598,809


##### handling exact duplicate values

In [4]:
# check duplicates
duplicate_transactions = transactions.duplicated().sum()

print("Number of Exact Duplicate Rows: ")
print(duplicate_transactions)

# inspect duplicates
display(transactions[transactions.duplicated(keep=False)])

Number of Exact Duplicate Rows: 
185


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
0,60567026-f719-4cd6-849e-137e86d8938f,5ff75116-0a50-4d04-80fb-31e5ccbb0769,2024-05-15,117.64,Clothing,PayPal
9,474f9233-0616-431b-b3e9-a6feabe68abb,f8bde21a-7a6b-41f6-ac39-eeb77640fa9b,2023-06-15,149.39,Clothing,Debit Card
16,54b3556e-e480-41a2-9dcd-7da1d91655aa,c181a599-fcc7-4f9b-b618-783684e0fd24,2023-03-06,861.39,Clothing,Bank Transfer
36,9560ed71-7a04-4cd9-a576-b622c5690487,ad529898-c1b5-4fe7-b30b-a9f19c666386,2024-05-31,833.48,Automotive,Bank Transfer
45,d3493edf-9847-46ef-b400-7063b54bcf63,6fa36483-8ac7-4af8-9cc7-e0607673f331,2022-07-15,119.65,Clothing,PayPal
...,...,...,...,...,...,...
3195,59f8afe5-173b-4e52-b641-f98183f1a891,3121df10-58e5-4ed4-a05c-b1f12df6bcc7,2023-01-18,892.44,Home & Garden,Bank Transfer
3196,16d0b561-279d-4da9-9aa8-8c5906dff79a,dff4f636-2c55-4237-8c55-efb15b05949b,2024-02-13,431.64,Health & Beauty,Credit Card
3197,34e3f558-622e-4d6b-8d5d-07ab632e1b95,db0a4640-ea96-4fcd-989b-b71ba767529c,2022-12-26,NaN,Home & Garden,PayPal
3198,d83178cf-653d-435c-a7e8-880361dec6e7,89a1c0bc-4d49-4913-bb0f-26bd56b46a34,2022-09-26,343.46,Health & Beauty,Debit Card


In [5]:
# remove duplicates
transaction_cleaned = transactions.drop_duplicates().copy()

# verify
print("Duplicates Remaining:")
print(transaction_cleaned.duplicated().sum())

Duplicates Remaining:
0


#### handling missing and invalid values

##### amount column

In [8]:
# check missing values
missing_values = transaction_cleaned.isnull().sum()
print("Missing Values:")
print(missing_values)

Missing Values:
CustomerID           0
TransactionID        0
TransactionDate      0
Amount             283
ProductCategory    282
PaymentMethod        0
dtype: int64


In [9]:
# inspect data types and check numerical statistics
print("Data Types:")
print(transaction_cleaned.dtypes)

print("\nNumerical Statistics:")
display(transaction_cleaned.describe())

Data Types:
CustomerID         str
TransactionID      str
TransactionDate    str
Amount             str
ProductCategory    str
PaymentMethod      str
dtype: object

Numerical Statistics:


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
count,3015,3015,3015,2732,2733,3015
unique,1871,3000,808,2603,5,4
top,e54d1219-b6b3-4b5c-8b75-62f4148c23b6,5f93f1dc-0363-4e69-979b-f5d58fd42658,2022-10-31,-100,Clothing,Debit Card
freq,6,2,11,39,571,767


In [12]:
# check categorical inconsistencies
for column in transaction_cleaned.select_dtypes(include="object").columns:
    print(f"\n{column}")
    print(transaction_cleaned[column].value_counts(dropna=False))


CustomerID
CustomerID
e54d1219-b6b3-4b5c-8b75-62f4148c23b6    6
49f06ea9-4ae5-4ce5-98ce-d2c2f6f15cbd    6
b03edff3-4a64-4e5e-8818-10e37291a8c6    6
577fcc56-95dc-443a-9289-d12614f5cea3    6
799943c7-3b71-4f24-9b16-c2fcaf9b3871    5
                                       ..
6fa9f01c-09f8-44fd-8e18-51ca4f3921bf    1
e4e86de7-3972-407e-9a7d-66b34a347463    1
7fa1d4e1-e465-4d23-82c8-c096f58fb5ed    1
167f3b4b-5ad6-4fe0-a51d-8f1c52cbee27    1
6c0b47c9-e467-433d-a0f1-9a6fe8adcd38    1
Name: count, Length: 1871, dtype: int64

TransactionID
TransactionID
5f93f1dc-0363-4e69-979b-f5d58fd42658    2
7202af3a-448f-4970-9674-40f736cfdfde    2
185ee140-348a-4461-bd40-07a7e8535815    2
27fb25aa-f148-423f-967c-4d1764e26b29    2
f39c8b1d-595a-4656-8108-25981bf3fa14    2
                                       ..
64b60afb-2c36-45be-b62a-7f471fef2077    1
7961f74a-b3ea-49e3-aa13-e55f3ef41d71    1
5e0e3958-fcf9-40ed-9a90-0f0920c244aa    1
b4a98e4d-c0ab-4282-99a9-3d909befd473    1
8c2bbc45-1053-42de-9310-b8

C:\Users\jrbusadre\AppData\Local\Temp\ipykernel_27964\2982883701.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for column in transaction_cleaned.select_dtypes(include="object").columns:


##### amount column

In [13]:
# convert amount from string to numeric

transaction_cleaned["Amount"] = pd.to_numeric(
    transaction_cleaned["Amount"],
    errors="coerce"
)

# check numerical statistics

print("Numerical Statistics of Amount Column:")
display(transaction_cleaned["Amount"].describe())

# check again number of missing values

print("Number of Missing Values:")
print(transaction_cleaned["Amount"].isnull().sum())

Numerical Statistics of Amount Column:


count    2702.000000
mean      489.788919
std       299.239453
min      -100.000000
25%       228.240000
50%       496.625000
75%       746.697500
max       999.860000
Name: Amount, dtype: float64

Number of Missing Values:
313


### invalid amount values are dealt with prior to filling or dropping missing values

In [14]:
# count how many invalid amount values are in the dataset

invalid_amount = transaction_cleaned[
    (transaction_cleaned["Amount"] < 0) |
    (transaction_cleaned["Amount"] == 0)
]

print("Invalid Amount Values:")
print(len(invalid_amount))

display(invalid_amount)

Invalid Amount Values:
70


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
134,b94a126b-755e-4ee1-91f6-c5f86ceef780,793af848-8138-40e7-929c-67254ea56ca7,2023-07-22,0.0,Home & Garden,PayPal
196,89480374-82ec-44e5-ae61-bbed82bfd5b1,a6b7182d-6c74-4fa5-9491-035e49e5eb87,2023-11-09,0.0,Electronics,Bank Transfer
274,e6f4ae74-d58d-4770-8ebc-89c1f8778a6d,814d25d8-690b-45bb-96c9-b7a3e92a61db,2023-01-25,-100.0,NaN,PayPal
308,343ef45e-c257-47f1-bbd2-5683734d2fcd,c0ea759c-a09f-4989-bc3b-f56a0c0400d8,2022-08-13,0.0,Home & Garden,PayPal
319,b91999f9-b7ed-4da4-a285-1dafae98e21d,fb908e24-4c1e-4e24-a1b0-62465fa0515c,2024-03-22,-100.0,Health & Beauty,Debit Card
...,...,...,...,...,...,...
3050,a9af986b-d677-421a-81df-3d289ce9e38f,5f93f1dc-0363-4e69-979b-f5d58fd42658,2023-09-30,-100.0,NaN,Bank Transfer
3073,e7cf6999-b4ec-4ce4-bb09-bfd18d61d933,2e0b66a9-af5d-4cf9-bb60-55ab2932edee,2023-05-10,-100.0,Electronics,Bank Transfer
3123,8de3d5b9-1f1a-4bd7-b242-6cb09d7f3609,5fca256b-cd5a-48fa-8b07-37f39adde8c9,2023-05-17,-100.0,Clothing,Bank Transfer
3174,6a84f1a7-f636-4300-bd8d-c7f888460829,185ee140-348a-4461-bd40-07a7e8535815,2023-10-09,0.0,Health & Beauty,PayPal


In [15]:
# convert invalid amount values to missing values

transaction_cleaned.loc[
    (transaction_cleaned["Amount"] < 0) |
    (transaction_cleaned["Amount"] == 0),
    "Amount"
] = pd.NA

# validate

print("Missing Amount Values:")
print(transaction_cleaned["Amount"].isnull().sum())

print("\nAmount Statistics After Removing Invalid Values:")
display(transaction_cleaned["Amount"].describe())

Missing Amount Values:
383

Amount Statistics After Removing Invalid Values:


count    2632.000000
mean      504.334977
std       289.294142
min        10.030000
25%       245.260000
50%       508.070000
75%       750.420000
max       999.860000
Name: Amount, dtype: float64

In [16]:
# fill missing amount values with the median

transaction_cleaned["Amount"] = transaction_cleaned["Amount"].fillna(
    transaction_cleaned["Amount"].median()
)

# verify

print("Missing Amount Values After Filling:")
print(transaction_cleaned["Amount"].isnull().sum())

print("\nAmount Statistics After Cleaning:")
display(transaction_cleaned["Amount"].describe())

Missing Amount Values After Filling:
0

Amount Statistics After Cleaning:


count    3015.000000
mean      504.809443
std       270.291905
min        10.030000
25%       282.020000
50%       508.070000
75%       720.395000
max       999.860000
Name: Amount, dtype: float64

##### product category column

In [17]:
# inspect missing values

print("Product Category Values:")
display(transaction_cleaned["ProductCategory"].value_counts(dropna=False))

Product Category Values:


ProductCategory
Clothing           571
Electronics        564
Automotive         540
Home & Garden      530
Health & Beauty    528
NaN                282
Name: count, dtype: int64

#### missing product category values are replaced with 'Unknown' because the original category is unavailable

In [18]:
# fill missing values with 'Unknown'

transaction_cleaned["ProductCategory"] = transaction_cleaned[
    "ProductCategory"
].fillna("Unknown")

# verify

print("Product Category Values After Cleaning:")
display(transaction_cleaned["ProductCategory"].value_counts(dropna=False))

Product Category Values After Cleaning:


ProductCategory
Clothing           571
Electronics        564
Automotive         540
Home & Garden      530
Health & Beauty    528
Unknown            282
Name: count, dtype: int64

##### transaction date column

In [19]:
# convert transaction date to datetime

transaction_cleaned["TransactionDate"] = pd.to_datetime(
    transaction_cleaned["TransactionDate"],
    errors="coerce",
    format="mixed",
    dayfirst=True
)

# verify

print("Transaction Date Data Type:")
print(transaction_cleaned["TransactionDate"].dtype)

Transaction Date Data Type:
datetime64[us]


In [20]:
# check missing or invalid transaction dates

print("Missing Transaction Dates:")
print(transaction_cleaned["TransactionDate"].isnull().sum())

print("\nTransaction Date Range:")
print("Minimum Date:", transaction_cleaned["TransactionDate"].min())
print("Maximum Date:", transaction_cleaned["TransactionDate"].max())

Missing Transaction Dates:
0

Transaction Date Range:
Minimum Date: 2022-01-07 00:00:00
Maximum Date: 2024-12-06 00:00:00


##### payment method column

In [21]:
# inspect payment methods

print("Payment Method Values:")
display(transaction_cleaned["PaymentMethod"].value_counts(dropna=False))

Payment Method Values:


PaymentMethod
Debit Card       767
Credit Card      755
PayPal           750
Bank Transfer    743
Name: count, dtype: int64

##### customer id column

In [22]:
# investigate duplicate customer id values

duplicate_customer_ids = transaction_cleaned[
    transaction_cleaned.duplicated(
        subset="CustomerID",
        keep=False
    )
]

print("Duplicate Customer ID Records:")
print(len(duplicate_customer_ids))

display(duplicate_customer_ids.head(20))

Duplicate Customer ID Records:
1959


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
1,4090ba85-b111-4f75-a792-c777965f5255,2c39b9fe-ff57-4d39-9321-9f5cdf187aa1,2023-04-26,466.14,Health & Beauty,Bank Transfer
4,6e3e8eb8-bc0f-4ffe-9f74-5d5efec9502f,2034aebc-8280-4254-a667-92bcd1c2be4f,2024-03-06,590.52,Home & Garden,Bank Transfer
6,7fbc5847-a652-4fc8-a7d8-5e85d846e91b,591f1ea2-c97d-44dd-9539-71ac89b94ff3,2024-12-01,508.07,Automotive,PayPal
7,958e5c8d-48ca-42dd-bb71-a766a374233a,fb24e098-3ab9-40a2-bcc3-b8ebb23f549a,2023-10-03,399.70,Home & Garden,PayPal
8,39c6e7d2-6c4b-44c0-8961-ddc1ecbdb0c6,833b026a-7c02-4101-832d-62c07569b0f6,2024-01-26,296.99,Clothing,PayPal
9,474f9233-0616-431b-b3e9-a6feabe68abb,f8bde21a-7a6b-41f6-ac39-eeb77640fa9b,2023-06-15,149.39,Clothing,Debit Card
10,187ec535-ad79-418a-aa08-86fd8a2becbb,85dc5e9e-1bd7-44e3-9a13-980736fedd3c,2023-01-06,255.41,Health & Beauty,Bank Transfer
11,848f7178-b94b-4d8e-8baa-6a3886828887,80cd99c4-01c7-47ec-8272-79250c1c4725,2024-09-03,813.81,Electronics,Credit Card
12,c3a5e86e-a74b-420e-b1e7-13bcbe6b71ac,e101f309-f9f8-4856-aa30-0153aa5f368c,2022-06-11,989.99,Electronics,Debit Card
14,28c29c18-0cbe-40ea-b218-a96430cc4409,deaec587-bd63-4556-8cea-e9d297f6651e,2023-08-11,140.88,Automotive,Credit Card


##### transaction id column

In [23]:
# investigate duplicate transaction id values

duplicate_transaction_ids = transaction_cleaned[
    transaction_cleaned.duplicated(
        subset="TransactionID",
        keep=False
    )
]

print("Duplicate Transaction ID Records:")
print(len(duplicate_transaction_ids))

display(
    duplicate_transaction_ids.sort_values("TransactionID")
)

Duplicate Transaction ID Records:
30


,CustomerID,TransactionID,TransactionDate,Amount,ProductCategory,PaymentMethod
2118,778da54b-8f96-4609-a21b-bba3d7f071ea,022cfd97-a8e8-4e50-8700-2720d52bfd75,2022-08-23,586.87,Health & Beauty,PayPal
3005,778da54b-8f96-4609-a21b-bba3d7f071ea,022cfd97-a8e8-4e50-8700-2720d52bfd75,2022-08-23,508.07,Health & Beauty,PayPal
413,6a84f1a7-f636-4300-bd8d-c7f888460829,185ee140-348a-4461-bd40-07a7e8535815,2023-09-10,953.86,Health & Beauty,PayPal
3174,6a84f1a7-f636-4300-bd8d-c7f888460829,185ee140-348a-4461-bd40-07a7e8535815,2023-09-10,508.07,Health & Beauty,PayPal
431,b94b3531-591d-445c-b282-dcacb0baca5b,27fb25aa-f148-423f-967c-4d1764e26b29,2022-07-30,506.96,Automotive,Credit Card
3038,b94b3531-591d-445c-b282-dcacb0baca5b,27fb25aa-f148-423f-967c-4d1764e26b29,2024-04-18,508.07,Automotive,Credit Card
2081,e7cf6999-b4ec-4ce4-bb09-bfd18d61d933,2e0b66a9-af5d-4cf9-bb60-55ab2932edee,2023-10-05,504.42,Electronics,Bank Transfer
3073,e7cf6999-b4ec-4ce4-bb09-bfd18d61d933,2e0b66a9-af5d-4cf9-bb60-55ab2932edee,2023-10-05,508.07,Electronics,Bank Transfer
220,a9af986b-d677-421a-81df-3d289ce9e38f,5f93f1dc-0363-4e69-979b-f5d58fd42658,2023-09-30,125.77,Unknown,Bank Transfer
3050,a9af986b-d677-421a-81df-3d289ce9e38f,5f93f1dc-0363-4e69-979b-f5d58fd42658,2023-09-30,508.07,Unknown,Bank Transfer


#### final dataset verification

In [24]:
# verify cleaned dataset

print("Cleaned Dataset Information:")
transaction_cleaned.info()

print("\nCleaned Dataset Dimensions:")
print(transaction_cleaned.shape)

print("\nRemaining Missing Values:")
display(transaction_cleaned.isnull().sum())

print("\nRemaining Exact Duplicates:")
print(transaction_cleaned.duplicated().sum())

Cleaned Dataset Information:
<class 'pandas.DataFrame'>
Index: 3015 entries, 0 to 3177
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   CustomerID       3015 non-null   str           
 1   TransactionID    3015 non-null   str           
 2   TransactionDate  3015 non-null   datetime64[us]
 3   Amount           3015 non-null   float64       
 4   ProductCategory  3015 non-null   str           
 5   PaymentMethod    3015 non-null   str           
dtypes: datetime64[us](1), float64(1), str(4)
memory usage: 439.6 KB

Cleaned Dataset Dimensions:
(3015, 6)

Remaining Missing Values:


CustomerID         0
TransactionID      0
TransactionDate    0
Amount             0
ProductCategory    0
PaymentMethod      0
dtype: int64


Remaining Exact Duplicates:
0


### Important Note

> **The customer transactions data underwent preprocessing to improve its quality and consistency. Exact duplicate records were removed, invalid and missing Amount values were addressed, missing ProductCategory entries were assigned as "Unknown," and TransactionDate was converted into a standardized date format. After cleaning, the dataset contained 3,015 records across 6 columns, with no remaining missing values or exact duplicate rows.**
>
> **However, duplicate CustomerID values were retained because a single customer can have multiple transactions. Duplicate TransactionID values were examined but were not removed automatically, since further verification is necessary to determine whether they represent genuine records or duplicate transactions.**

In [26]:
# save cleaned dataset

transaction_cleaned.to_csv(
    "customer_transactions_preprocessed.csv",
    index=False
)